### ISSUE TITLE
file.append should be stateful, changes spike minion CPU to 100%

### ISSUE THREAD

**Issue by cmclaughlin** [saltstack/salt#31278]
I expected file.append to be stateful.  By that I mean if you append X to a file and later change X to Y, X should be removed from the file and Y should be added.  I found that when I change X and try to apply the highstate, it never returns and salt-minion process spikes to 100% until I restart it.

Here's my use-case:

supervisord is installed via yum and ships with some default config files.  I thought appending to one of the default config files was very elegant: 

	/etc/sysconfig/supervisord:
	  file.append:
	    - source: salt://mystate/files/sysconfig

However, a few weeks after the initial setup I found a need to add some values to the file.  I think that's a typical pattern which should be supported and should not pin a CPU or cause highstate to fail.

When the minion CPU spikes to 100% there are two salt-minion processes.  When I strace the one that's pegging the CPU, I don't see anything.  When I strace the other, I see lots of this:

	clock_gettime(CLOCK_MONOTONIC, {68298, 676974506}) = 0
	gettimeofday({1455734102, 441523}, NULL) = 0
	gettimeofday({1455734102, 441770}, NULL) = 0
	gettimeofday({1455734102, 442013}, NULL) = 0
	gettimeofday({1455734102, 442228}, NULL) = 0
	clock_gettime(CLOCK_MONOTONIC, {68298, 677977944}) = 0
	poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=24, events=POLLIN}], 3, 0) = 0 (Timeout)
	poll([{fd=16, events=POLLIN}], 1, 0)    = 0 (Timeout)
	poll([{fd=24, events=POLLIN}], 1, 0)    = 0 (Timeout)
	clock_gettime(CLOCK_MONOTONIC, {68298, 678454749}) = 0
	poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=24, events=POLLIN}], 3, 997) = 0 (Timeout)
	poll([{fd=16, events=POLLIN}], 1, 0)    = 0 (Timeout)
	poll([{fd=24, events=POLLIN}], 1, 0)    = 0 (Timeout)
	clock_gettime(CLOCK_MONOTONIC, {68299, 676855261}) = 0
	gettimeofday({1455734103, 441357}, NULL) = 0
	gettimeofday({1455734103, 441590}, NULL) = 0
	gettimeofday({1455734103, 441811}, NULL) = 0
	gettimeofday({1455734103, 441986}, NULL) = 0
	clock_gettime(CLOCK_MONOTONIC, {68299, 677738964}) = 0
	poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=24, events=POLLIN}], 3, 0) = 0 (Timeout)
	poll([{fd=16, events=POLLIN}], 1, 0)    = 0 (Timeout)
	poll([{fd=24, events=POLLIN}], 1, 0)    = 0 (Timeout)
	clock_gettime(CLOCK_MONOTONIC, {68299, 678096784}) = 0
	poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=24, events=POLLIN}], 3, 997) = 0 (Timeout)
	poll([{fd=16, events=POLLIN}], 1, 0)    = 0 (Timeout)
	poll([{fd=24, events=POLLIN}], 1, 0)    = 0 (Timeout)
	clock_gettime(CLOCK_MONOTONIC, {68300, 676357743}) = 0
	gettimeofday({1455734104, 440899}, NULL) = 0
	gettimeofday({1455734104, 441066}, NULL) = 0
	gettimeofday({1455734104, 441307}, NULL) = 0
	gettimeofday({1455734104, 441498}, NULL) = 0
	clock_gettime(CLOCK_MONOTONIC, {68300, 677246722}) = 0
	poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=24, events=POLLIN}], 3, 0) = 0 (Timeout)
	poll([{fd=16, events=POLLIN}], 1, 0)    = 0 (Timeout)
	poll([{fd=24, events=POLLIN}], 1, 0)    = 0 (Timeout)
	clock_gettime(CLOCK_MONOTONIC, {68300, 677599173}) = 0

Here's some version info...

master:

	Salt Version:
	           Salt: 2015.8.3

	Dependency Versions:
	         Jinja2: 2.7.2
	       M2Crypto: Not Installed
	           Mako: 0.9.1
	         PyYAML: 3.10
	          PyZMQ: 14.0.1
	         Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
	           RAET: Not Installed
	        Tornado: 4.2.1
	            ZMQ: 4.0.4
	           cffi: Not Installed
	       cherrypy: Not Installed
	       dateutil: 1.5
	          gitdb: 0.5.4
	      gitpython: 0.3.2 RC1
	          ioflo: Not Installed
	        libnacl: Not Installed
	   msgpack-pure: Not Installed
	 msgpack-python: 0.3.0
	   mysql-python: 1.2.3
	      pycparser: Not Installed
	       pycrypto: 2.6.1
	         pygit2: Not Installed
	   python-gnupg: 0.3.6
	          smmap: 0.8.2
	        timelib: Not Installed

	System Versions:
	           dist: Ubuntu 14.04 trusty
	        machine: x86_64
	        release: 3.13.0-48-generic
	         system: Ubuntu 14.04 trusty

minion:

	    Salt Version:
	               Salt: 2015.8.5

	    Dependency Versions:
	             Jinja2: 2.7.2
	           M2Crypto: 0.21.1
	               Mako: Not Installed
	             PyYAML: 3.11
	              PyZMQ: 14.5.0
	             Python: 2.6.9 (unknown, Dec 17 2015, 01:08:55)
	               RAET: Not Installed
	            Tornado: 4.2.1
	                ZMQ: 4.0.5
	               cffi: Not Installed
	           cherrypy: Not Installed
	           dateutil: Not Installed
	              gitdb: Not Installed
	          gitpython: Not Installed
	              ioflo: Not Installed
	            libgit2: Not Installed
	            libnacl: Not Installed
	       msgpack-pure: Not Installed
	     msgpack-python: 0.4.6
	       mysql-python: 1.2.3
	          pycparser: Not Installed
	           pycrypto: 2.6.1
	             pygit2: Not Installed
	       python-gnupg: Not Installed
	              smmap: Not Installed
	            timelib: Not Installed

	    System Versions:
	               dist:
	            machine: x86_64
	            release: 4.1.10-17.31.amzn1.x86_64

I'm switching my code to `file.blockreplace`, so I have a workaround.  But I think this should be considered as a serious bug.

Please let me know if there's anything I can do to help.

Thank you.

**Comment by jfindlay** [comment]
@cmclaughlin, thanks for the report.  Statefully editing files can be tricky.  It's often simpler to have salt manage the whole file rather than pieces of it.  Like, for example, with the patch utility, or git, it is common for there to be conflicts that cannot be automatically resolved.

**Comment by cmclaughlin** [comment]
@jfindlay agreed... btw, I discussed this with some co-workers - they think I'm crazy and they don't think append should be stateful.  Either way, we should probably figure out the endless CPU spike.

**Comment by cachedout** [comment]
`file.append` should certainly be stateful, it's just sort of poorly named. It's real intention is more along the lines of what could be better named `file.endswith`. The strace you posted is just the minion idiling on the socket poll waiting for the state to finish. I'd like to see an strace of the other from the point the state starts running. You could potentially accomplish this with `salt-call`.

**Comment by cmclaughlin** [comment]
@cachedout I was able to strace the state... there's tons of output, but these are the last few lines.  

	munmap(0x7f469b970000, 4096)            = 0
	lstat("/etc", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
	lstat("/etc/sysconfig", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
	lstat("/etc/sysconfig/supervisord", {st_mode=S_IFREG|0640, st_size=3131, ...}) = 0
	open("/etc/sysconfig/supervisord", O_RDONLY) = 44
	fstat(44, {st_mode=S_IFREG|0640, st_size=3131, ...}) = 0
	fcntl(44, F_GETFD)                      = 0
	fcntl(44, F_SETFD, FD_CLOEXEC)          = 0
	fstat(44, {st_mode=S_IFREG|0640, st_size=3131, ...}) = 0
	dup(44)                                 = 45
	mmap(NULL, 3131, PROT_READ, MAP_SHARED, 44, 0) = 0x7f469b970000

Does this tell you anything?
Please let me know if you need more info.

**Comment by cachedout** [comment]
Wild guess here but does the CPU spike happen if you change this to append to a file other than that one?

**Comment by cmclaughlin** [comment]
btw, I have some other stuff that references that file... possibly contributing the CPU spike. 

In my supervisord state I manage the ownership and mode of that file:

	supervisord-sysconfig-mode:
	  file.managed:
	    - name: /etc/sysconfig/supervisord
	    - user: root
	    - group: supervisord
	    - mode: 0640

Every app that I configure to run under supervisord sets it's own sysconfig file, as shown above:

	/etc/sysconfig/supervisord:
	  file.append:
	    - source: salt://mystate/files/sysconfig

And after dropping a file in /etc/supervisord.d configuring supervisord to run an app, I configure a watch so that supervisord is restarted if the sysconfig file changes:

	sysconfig-watch:
	  file.managed:
	    - name: /etc/sysconfig/supervisord
	    - watch_in:
	      - service: supervisor-service

Sorry I didn't explain all this in detail from the beginning... I just remembered there's more to it.  I don't have much to go off, but I wonder if the watch contributes to my problem...?

**Comment by cachedout** [comment]
Ah-ha! Yeah, that sounds *way* more likely. Can you wiggle that stuff around and see if the problem goes away? I'll bet it does.

**Comment by cmclaughlin** [comment]
Not sure if this is good news or bad news, but I removed all the other references to that file and was able to reproduce it.  The only thing I left is the file.append.  I'll keep poking around...

**Comment by cachedout** [comment]
That surprises me, to be honest. Can you try switching the file.append to a different file, just as a test?

**Comment by cmclaughlin** [comment]
I did some more testing...

**Comment by cmclaughlin** [comment]
I did some more testing... I think it just breaks down as a very complicated regex (assuming that's what's going on behind the scenes).

If I start from scratch and add a character to new line to file that is managed with file.append and run highstate... and keep repeating this pattern over and over again, I eventually hit a point where it completely breaks.

Leading up to that I see the CPU spiked out at times, but with relatively small files it returns successfully.  With this test I keep appending over and over again, so the file gets longer and more complicated.

Switching to a different file helps in the sense that the test gets reset.  But if I keep changing the source that's appended to the new file it should fail the same way once the file gets long or complicated enough.

**Comment by cachedout** [comment]
So, I re-read this issue today and I wanted to clarify something. It sounds like your expectation is that after setting a string to be appended to a file that the string should be _replaced_ if it's changed in the state. In this sense, I have to side with your co-workers I'm afraid. :] I don't believe this has ever been the intended behavior of the state, nor should it be in my view. To do so, would require that we track not only the current state of the file and the string to be appended, but the _previous_ string which the state was appending. In my view, that's well beyond the scope of what the state was ever intended to do. I think your inclination to use `file.blockreplace` for that workflow is far better.

I'm I'm misunderstanding your intended workflow, please let me know.

That said, I am troubleshooting this high-CPU report and I'll get back to you on that. I just wanted to mention this while I was thinking about it.

**Comment by cachedout** [comment]
Can you be a little more specific about what sort of performance you're seeing and what you're expecting? I'm running `file.append` state against a file that's 96MB and with 1.2M lines of random text and while the CPU spikes (as I would fully expect it to), it's only for about 1.2s.

I've tested this on the head of the 2015.8 branch and against v2015.8.3.

I'm not sure what else to do here to try and reproduce this so I'm setting this to `Cannot Reproduce` until we can get a specific reproduction test case for this. I think we'll need to see a state, a template and a target file that causes this problem in order to go forward here. Thanks.

**Comment by cachedout** [comment]
I've had no response on the questions I've asked here and have been unable to reproduce this. Therefore, I am setting this to closed. If some responses come back, we'll happily re-open this. Thanks.

**Comment by cmclaughlin** [comment]
@cachedout Sorry for the delay.  I just did some more testing...

10mb file with 129870 lines of text.  Applying highstate takes 162.689 seconds, but timing the salt CLI takes 2m55.386s.  Running on 2 core box, 1 core is maxed out.

Making a change to the appended text in the target file so it's appended again takes 156.410 s to apply the highstate and 2m37.997s for the command to return.  So maybe the slowness isn't the result of a complicated regex as I was previously thinking. 

This is the simple test I ran:

	/tmp/bootstrap-salt.log:
	  file.append:
	    - source: salt://charlestest/files/blob

The blob file was generated with `base64 /dev/urandom | head -c 10000000`.

Not a real world example, but since I stopped using file.append in favor of file.blockreplace this is what I slapped together to test. 

Regarding expectations... 1s like you're seeing seems fine.  I'm at SaltConf if you want to take a closer look together.  Maybe I'm doing something stupid or need to do some tuning.
